core/Pos/Core/Slotting.hs

```diff
@@ -71,6 +79,14 @@ crucialSlot epochIdx = SlotId {siEpoch = epochIdx - 1, ..}
         leftToPanic "crucialSlot: " $
         mkLocalSlotIndex (fromIntegral (epochSlots - slotSecurityParam - 1))
 
+instance HasEpochIndex EpochOrSlot where
+    epochIndexL = lens (epochOrSlot identity siEpoch) setter
```

Comment:
I think it's a mistake that `EpochOrSlot` is a newtype over `Either EpochIndex SlotId`. It should be `EpochOrSlot = EOSepoch !EpochIndex | EOSslot !SlotId`.